File: topos/graphs/mdg/object.py
"""
Module Dependency Graph (MDG) Representation
============================================
Consumes the knowledge graph produced by `GitNexus
<https://github.com/abhigyanpatwari/GitNexus>`_ and lifts it into a
:class:`~topos.graphs.base.Representation`.  This is the **inter-module**
view of the program — it captures the import/call/inheritance structure
across files, packages, and classes.  Compare this with the academic
**intra-procedural Program Dependence Graph** at
:mod:`topos.graphs.pdg.object`, which records control- and data-dependence
edges *within* a single procedure.

GitNexus runs ``gitnexus analyze`` on a repository and writes a
``.gitnexus/`` directory containing a LadybugDB graph store.  This module
parses that output into an in-memory graph of typed nodes and
relationships, then computes dependency-level metrics that the AST alone
cannot provide.

Node labels and relationship types mirror GitNexus's shared schema::

    Nodes:  File, Module, Function, Class, Method, Import, ...
    Edges:  CALLS, IMPORTS, INHERITS, CONTAINS, USES, ...

Metrics produced (feed the COMPOSABLE generator of H(G_qual)):
    - ``mdg.coupling``   -- afferent + efferent coupling for a file
    - ``mdg.instability`` -- Ce / (Ca + Ce)  (Martin's metric)
    - ``mdg.fan_in``      -- incoming CALLS edges
    - ``mdg.fan_out``     -- outgoing CALLS edges
    - ``mdg.dep_depth``   -- longest IMPORTS chain from the file

"""

from __future__ import annotations

import json
from collections import defaultdict
from dataclasses import dataclass, field
from pathlib import Path
from typing import Literal

NodeLabel = Literal[
    "Project",
    "Package",
    "Module",
    "Folder",
    "File",
    "Class",
    "Function",
    "Method",
    "Variable",
    "Interface",
    "Enum",
    "Decorator",
    "Import",
    "Type",
    "CodeElement",
    "Community",
    "Process",
    "Struct",
    "Namespace",
    "Trait",
    "Constructor",
]


class LadybugSchemaMismatchError(RuntimeError):
    """Raised when ``.gitnexus/lbug`` storage version exceeds embedded ladybug."""

    def __init__(self, message: str, *, original: Exception | None = None) -> None:
        super().__init__(message)
        self.original = original


class LadybugBranchMismatchError(FileNotFoundError):
    """Raised when ``.gitnexus`` has indexed stores, but none for the current branch.

    Subclasses ``FileNotFoundError`` (same trick ``LadybugSchemaMismatchError``
    plays on ``RuntimeError``) so every existing ``except FileNotFoundError``
    elsewhere in the codebase keeps degrading gracefully with no further
    changes -- this only needs special-casing where a *better* message/state
    is wanted, not everywhere graceful degradation already happens.
    """


def _is_shadow_replay_error(exc: BaseException) -> bool:
    """Whether *exc* is Ladybug refusing to replay shadow pages read-only.

    Incrementally-updated ``.gitnexus`` stores can be left with pending
    shadow pages (e.g. after ``gitnexus analyze`` without a full wipe).
    Ladybug requires a read-write handle to replay them.
    """
    msg = str(exc).lower()
    return "shadow page" in msg and "read-only" in msg


def _raise_schema_mismatch(exc: BaseException) -> None:
    msg = str(exc).lower()
    if "different version" not in msg and "storage version" not in msg:
        raise exc
    raise LadybugSchemaMismatchError(
        "LadybugDB storage version mismatch while loading .gitnexus/lbug. "
        "Upgrade Topos to v0.3.4+ (bundles ladybug 0.17+) or re-run "
        "'topos depgraph generate' after upgrading. "
        "GitNexus 1.6.x requires ladybug 0.17+.",
        original=exc if isinstance(exc, Exception) else None,
    ) from exc


RelationshipType = Literal[
    "CONTAINS",
    "CALLS",
    "INHERITS",
    "IMPORTS",
    "USES",
    "DEFINES",
    "DECORATES",
    "IMPLEMENTS",
    "EXTENDS",
    "HAS_METHOD",
    "HAS_PROPERTY",
    "ACCESSES",
    "MEMBER_OF",
    "METHOD_OVERRIDES",
    "METHOD_IMPLEMENTS",
    "STEP_IN_PROCESS",
]


@dataclass
class GraphNode:
    """A node in the GitNexus knowledge graph."""

    id: str
    label: str
    properties: dict[str, object] = field(default_factory=dict)


@dataclass
class GraphRelationship:
    """An edge in the GitNexus knowledge graph."""

    id: str
    source_id: str
    target_id: str
    type: str
    confidence: float = 1.0
    reason: str = ""
    properties: dict[str, object] = field(default_factory=dict)


def _parse_node(item: dict) -> GraphNode:
    return GraphNode(
        id=item["id"],
        label=item["label"],
        properties=item.get("properties", {}),
    )


def _parse_relationship(item: dict) -> GraphRelationship:
    return GraphRelationship(
        id=item.get("id", f"{item['sourceId']}->{item['targetId']}"),
        source_id=item["sourceId"],
        target_id=item["targetId"],
        type=item["type"],
        confidence=item.get("confidence", 1.0),
        reason=item.get("reason", ""),
        properties=item.get("properties", {}),
    )


@dataclass
class ModuleDependencyGraph:
    """
    Inter-module dependency-graph representation parsed from GitNexus output.

    Provides graph lookup methods and computes dependency-level
    metrics for a target file path within the graph.

    This is the **module-level** dependency view (imports, calls, inheritance
    across files).  It is distinct from the academic intra-procedural
    Program Dependence Graph at :mod:`topos.graphs.pdg.object`.

    Attributes:
        target_file: The file path to compute metrics for.
        nodes: All nodes in the graph, keyed by ID.
        relationships: All relationships in the graph, keyed by ID.
    """

    target_file: str
    nodes: dict[str, GraphNode] = field(default_factory=dict)
    relationships: dict[str, GraphRelationship] = field(default_factory=dict)

    _outgoing: dict[str, list[GraphRelationship]] = field(
        default_factory=lambda: defaultdict(list), repr=False, compare=False
    )
    _incoming: dict[str, list[GraphRelationship]] = field(
        default_factory=lambda: defaultdict(list), repr=False, compare=False
    )

    @property
    def name(self) -> str:
        return "mdg"

    @property
    def dimension(self) -> str:
        # Feeds the COMPOSABLE generator of H(G_qual).
        return "composable"

    # ------------------------------------------------------------------
    # Construction
    # ------------------------------------------------------------------

    @classmethod
    def from_gitnexus_dir(
        cls, gitnexus_dir: str | Path, target_file: str
    ) -> ModuleDependencyGraph:
        """
        Build a ModuleDependencyGraph from a ``.gitnexus/`` directory.

        Supports both the current LadybugDB binary format (``lbug`` file,
        produced by GitNexus ≥ 1.5) and the legacy JSON directory format.

        Args:
            gitnexus_dir: Path to the ``.gitnexus/`` directory.
            target_file: The file path to compute metrics for.

        Raises:
            FileNotFoundError: If the graph store cannot be found.
            ImportError: If ``ladybug`` is not installed for binary format.
            LadybugSchemaMismatchError: If the store version exceeds embedded ladybug.
        """
        from topos.utils.gitnexus import current_git_branch, resolve_lbug_store

        base = Path(gitnexus_dir)
        branch = current_git_branch(base.parent)
        resolved = resolve_lbug_store(base, branch)
        lbug_path = resolved.path

        if lbug_path is None:
            if resolved.available_branches:
                raise LadybugBranchMismatchError(
                    f"No GitNexus store indexed for branch {branch!r} at {base}. "
                    f"Indexed branches: {', '.join(resolved.available_branches)}. "
                    "Run 'gitnexus analyze' on this branch (or 'topos depgraph "
                    "generate') to index it."
                )
            raise FileNotFoundError(
                f"LadybugDB store not found at {base / 'lbug'}. "
                "Install GitNexus (pnpm add -g gitnexus, or npm install -g gitnexus) and run "
                "'gitnexus analyze' in the repository root first."
            )

        if lbug_path.is_file():
            return cls._from_ladybugdb(lbug_path, target_file)

        if lbug_path.is_dir():
            return cls._from_json_dir(lbug_path, target_file)

        raise FileNotFoundError(
            f"LadybugDB store not found at {lbug_path}. "
            "Install GitNexus (pnpm add -g gitnexus, or npm install -g gitnexus) and run "
            "'gitnexus analyze' in the repository root first."
        )

    @classmethod
    def _from_ladybugdb(
        cls, lbug_path: Path, target_file: str
    ) -> ModuleDependencyGraph:
        """Load from the binary LadybugDB format produced by GitNexus ≥ 1.5."""
        import ladybug as lb

        graph = cls(target_file=target_file)
        try:
            db = lb.Database(str(lbug_path), read_only=True)
        except RuntimeError as exc:
            if not _is_shadow_replay_error(exc):
                _raise_schema_mismatch(exc)
            # Pending shadow pages (e.g. incremental `gitnexus analyze` without
            # a full wipe) can only be replayed with a read-write handle.
            try:
                db = lb.Database(str(lbug_path), read_only=False)
            except RuntimeError as retry_exc:
                _raise_schema_mismatch(retry_exc)
        conn = lb.Connection(db)

        # Discover node tables at runtime so we're not tied to a fixed schema.
        tables_result = conn.execute("CALL show_tables() RETURN *")
        node_tables = []
        while tables_result.has_next():
            row = tables_result.get_next()
            # row: [id, name, type, database, comment]
            if len(row) >= 3 and row[2] == "NODE":
                node_tables.append(row[1])

        for label in node_tables:
            result = conn.execute(f"MATCH (n:`{label}`) RETURN n")
            while result.has_next():
                (node_data,) = result.get_next()
                node_id = node_data.get("id")
                if node_id is None:
                    continue
                props = {k: v for k, v in node_data.items() if not k.startswith("_")}
                graph.add_node(GraphNode(id=node_id, label=label, properties=props))

        # Load all relationships from the single CodeRelation table. Also try
        # to pull `step` (the 1-indexed STEP_IN_PROCESS ordering property, see
        # topos/graphs/process/object.py) — not every ladybug schema version
        # carries this column on CodeRelation, so fall back to the plain query
        # if it isn't there rather than failing the whole load.
        has_step = True
        try:
            result = conn.execute(
                "MATCH (src)-[r:CodeRelation]->(dst) "
                "RETURN src.id, dst.id, r.type, r.confidence, r.reason, r.step"
            )
        except RuntimeError:
            has_step = False
            result = conn.execute(
                "MATCH (src)-[r:CodeRelation]->(dst) "
                "RETURN src.id, dst.id, r.type, r.confidence, r.reason"
            )
        idx = 0
        while result.has_next():
            row = result.get_next()
            if has_step:
                src_id, dst_id, rel_type, confidence, reason, step = row
            else:
                src_id, dst_id, rel_type, confidence, reason = row
                step = None
            if src_id is None or dst_id is None:
                continue
            graph.add_relationship(
                GraphRelationship(
                    id=f"{src_id}->{dst_id}:{rel_type}:{idx}",
                    source_id=src_id,
                    target_id=dst_id,
                    type=rel_type,
                    confidence=confidence or 1.0,
                    reason=reason or "",
                    properties={"step": step} if step is not None else {},
                )
            )
            idx += 1

        return graph

    @classmethod
    def _from_json_dir(cls, lbug_dir: Path, target_file: str) -> ModuleDependencyGraph:
        """Load from the legacy JSON directory format produced by GitNexus < 1.5."""
        graph = cls(target_file=target_file)
        for json_path in lbug_dir.glob("*.json"):
            data = json.loads(json_path.read_text(encoding="utf-8"))
            if isinstance(data, list):
                for item in data:
                    if "label" in item and "id" in item:
                        graph.add_node(_parse_node(item))
                    elif "type" in item and "sourceId" in item:
                        graph.add_relationship(_parse_relationship(item))
            elif isinstance(data, dict):
                for node_data in data.get("nodes", []):
                    graph.add_node(_parse_node(node_data))
                for rel_data in data.get("relationships", []):
                    graph.add_relationship(_parse_relationship(rel_data))
        return graph

    def add_node(self, node: GraphNode) -> None:
        self.nodes[node.id] = node

    def add_relationship(self, rel: GraphRelationship) -> None:
        self.relationships[rel.id] = rel
        self._outgoing[rel.source_id].append(rel)
        self._incoming[rel.target_id].append(rel)

    # ------------------------------------------------------------------
    # Lookups
    # ------------------------------------------------------------------

    def get_node(self, node_id: str) -> GraphNode | None:
        return self.nodes.get(node_id)

    def nodes_of_label(self, label: NodeLabel) -> list[GraphNode]:
        return [n for n in self.nodes.values() if n.label == label]

    def relationships_of_type(
        self, rel_type: RelationshipType
    ) -> list[GraphRelationship]:
        return [r for r in self.relationships.values() if r.type == rel_type]

    def outgoing(
        self, node_id: str, rel_type: RelationshipType | None = None
    ) -> list[GraphRelationship]:
        rels = self._outgoing.get(node_id, [])
        if rel_type is not None:
            return [r for r in rels if r.type == rel_type]
        return list(rels)

    def incoming(
        self, node_id: str, rel_type: RelationshipType | None = None
    ) -> list[GraphRelationship]:
        rels = self._incoming.get(node_id, [])
        if rel_type is not None:
            return [r for r in rels if r.type == rel_type]
        return list(rels)

    def file_node_id(self) -> str | None:
        """Find the File node ID matching ``target_file``."""
        for node in self.nodes.values():
            if node.label == "File":
                file_path = node.properties.get("filePath", "")
                if isinstance(file_path, str) and (
                    file_path == self.target_file
                    or file_path.endswith(f"/{self.target_file}")
                    or self.target_file.endswith(f"/{file_path}")
                ):
                    return node.id
        return None

    def contained_symbols(self, file_node_id: str) -> list[str]:
        """Return IDs of all symbols directly contained in a file node."""
        return [r.target_id for r in self.outgoing(file_node_id, "CONTAINS")]

    def all_contained_symbols(self, node_id: str) -> list[str]:
        """Return IDs of all symbols transitively reachable via CONTAINS edges.

        Performs a BFS down the CONTAINS tree starting from *node_id*.
        Cycles are handled safely via a visited set.
        """
        from collections import deque

        visited: set[str] = set()
        result: list[str] = []
        frontier: deque[str] = deque(self.contained_symbols(node_id))
        while frontier:
            child = frontier.popleft()
            if child in visited:
                continue
            visited.add(child)
            result.append(child)
            frontier.extend(self.contained_symbols(child))
        return result

    # ------------------------------------------------------------------
    # Metrics
    # ------------------------------------------------------------------

    def metrics(self) -> dict[str, float]:
        from topos.functors.probes.mdg.coupling import (
            calculate_coupling,
            calculate_dependency_depth,
            calculate_instability_from_result,
        )
        from topos.functors.probes.mdg.fan import calculate_fan_in_out

        file_id = self.file_node_id()
        if file_id is None:
            return {
                "mdg.coupling": 0.0,
                "mdg.instability": 0.5,
                "mdg.fan_in": 0.0,
                "mdg.fan_out": 0.0,
                "mdg.dep_depth": 0.0,
            }

        symbol_ids = set(self.all_contained_symbols(file_id))
        symbol_ids.add(file_id)
        coupling_result = calculate_coupling(self, file_id, symbol_ids)
        instability = calculate_instability_from_result(coupling_result)
        fan_result = calculate_fan_in_out(self, file_id, symbol_ids)
        dep_depth = calculate_dependency_depth(self, file_id)

        return {
            "mdg.coupling": float(coupling_result.total),
            "mdg.instability": instability,
            "mdg.fan_in": float(fan_result.fan_in),
            "mdg.fan_out": float(fan_result.fan_out),
            "mdg.dep_depth": float(dep_depth),
        }

File: topos/engine/src/graphs/mdg/object.rs
//! Module Dependency Graph (MDG) representation.
//!
//! Consumes the knowledge graph produced by [GitNexus](https://github.com/abhigyanpatwari/GitNexus)
//! and lifts it into a [`Representation`]. This is the **inter-module**
//! view of the program — it captures the import/call/inheritance
//! structure across files, packages, and classes. Compare this with the
//! academic **intra-procedural Program Dependence Graph** at
//! [`crate::graphs::pdg`], which records control- and data-dependence
//! edges *within* a single procedure.
//!
//! GitNexus runs `gitnexus analyze` on a repository and writes a
//! `.gitnexus/` directory containing a LadybugDB graph store. This
//! module parses that output into an in-memory graph of typed nodes and
//! relationships, then computes dependency-level metrics that the AST
//! alone cannot provide.
//!
//! Metrics produced (feed the COMPOSABLE generator of `H(G_qual)`):
//! - `mdg.coupling` — afferent + efferent coupling for a file
//! - `mdg.instability` — `Ce / (Ca + Ce)` (Martin's metric)
//! - `mdg.fan_in` — incoming CALLS edges
//! - `mdg.fan_out` — outgoing CALLS edges
//! - `mdg.dep_depth` — longest IMPORTS chain from the file
//!
//! # Deviation from the Python original: the LadybugDB binary format
//!
//! Python supports two `.gitnexus/` store formats: the legacy JSON
//! directory (ported here, [`ModuleDependencyGraph::from_json_dir`]) and
//! the current LadybugDB binary format (`lbug` file, GitNexus ≥ 1.5),
//! read via `import ladybug` — a Python-only binding to a native graph
//! database with no Rust client available. That's a genuine external
//! dependency wall, not a port left undone: reading it from Rust would
//! mean either an FFI binding to `ladybug`'s native library or
//! reimplementing a LadybugDB reader from scratch, neither of which
//! belongs in this representation module. [`MdgError::LadybugBinaryUnsupported`]
//! surfaces this clearly rather than silently returning an empty graph.

use super::models::{GraphNode, GraphRelationship};
use crate::functors::probes::mdg::coupling::{
    calculate_coupling, calculate_dependency_depth, calculate_instability_from_result,
};
use crate::functors::probes::mdg::fan::calculate_fan_in_out;
use crate::graphs::base::Representation;
use std::collections::{HashMap, VecDeque};
use std::path::PathBuf;

#[derive(Debug)]
pub enum MdgError {
    /// No `.gitnexus/lbug` store found at the expected path.
    NotFound(PathBuf),
    /// The store is the LadybugDB binary format — see the module doc.
    LadybugBinaryUnsupported(PathBuf),
    Io(std::io::Error),
    Json(serde_json::Error),
}

impl std::fmt::Display for MdgError {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            MdgError::NotFound(path) => write!(
                f,
                "LadybugDB store not found at {}. Install GitNexus (npm install -g gitnexus) \
                 and run 'gitnexus analyze' in the repository root first.",
                path.display()
            ),
            MdgError::LadybugBinaryUnsupported(path) => write!(
                f,
                "{} is the LadybugDB binary format, which topos-core cannot read \
                 (no Rust client for LadybugDB exists yet — see this module's doc comment)",
                path.display()
            ),
            MdgError::Io(e) => write!(f, "{e}"),
            MdgError::Json(e) => write!(f, "{e}"),
        }
    }
}

impl std::error::Error for MdgError {}

/// Inter-module dependency-graph representation parsed from GitNexus
/// output.
///
/// Provides graph lookup methods and computes dependency-level metrics
/// for a target file path within the graph. This is the **module-level**
/// dependency view (imports, calls, inheritance across files) — distinct
/// from the academic intra-procedural [`crate::graphs::pdg::object::ProgramDependenceGraph`].
#[derive(Debug, Clone)]
pub struct ModuleDependencyGraph {
    pub target_file: String,
    pub nodes: HashMap<String, GraphNode>,
    pub relationships: HashMap<String, GraphRelationship>,
    /// Relationship ids leaving each node — stores ids rather than
    /// cloned `GraphRelationship`s (Python shares object refs; this
    /// avoids the Rust equivalent of duplicating every relationship
    /// into two extra indices).
    outgoing: HashMap<String, Vec<String>>,
    incoming: HashMap<String, Vec<String>>,
}

impl ModuleDependencyGraph {
    pub fn new(target_file: impl Into<String>) -> Self {
        ModuleDependencyGraph {
            target_file: target_file.into(),
            nodes: HashMap::new(),
            relationships: HashMap::new(),
            outgoing: HashMap::new(),
            incoming: HashMap::new(),
        }
    }

    pub fn add_node(&mut self, node: GraphNode) {
        self.nodes.insert(node.id.clone(), node);
    }

    pub fn add_relationship(&mut self, rel: GraphRelationship) {
        self.outgoing
            .entry(rel.source_id.clone())
            .or_default()
            .push(rel.id.clone());
        self.incoming
            .entry(rel.target_id.clone())
            .or_default()
            .push(rel.id.clone());
        self.relationships.insert(rel.id.clone(), rel);
    }

    // --- Lookups -------------------------------------------------------

    pub fn get_node(&self, node_id: &str) -> Option<&GraphNode> {
        self.nodes.get(node_id)
    }

    pub fn nodes_of_label(&self, label: &str) -> Vec<&GraphNode> {
        self.nodes.values().filter(|n| n.label == label).collect()
    }

    pub fn relationships_of_type(&self, rel_type: &str) -> Vec<&GraphRelationship> {
        self.relationships
            .values()
            .filter(|r| r.rel_type == rel_type)
            .collect()
    }

    pub fn outgoing(&self, node_id: &str, rel_type: Option<&str>) -> Vec<&GraphRelationship> {
        self.outgoing
            .get(node_id)
            .into_iter()
            .flatten()
            .filter_map(|id| self.relationships.get(id))
            .filter(|r| rel_type.is_none_or(|t| r.rel_type == t))
            .collect()
    }

    pub fn incoming(&self, node_id: &str, rel_type: Option<&str>) -> Vec<&GraphRelationship> {
        self.incoming
            .get(node_id)
            .into_iter()
            .flatten()
            .filter_map(|id| self.relationships.get(id))
            .filter(|r| rel_type.is_none_or(|t| r.rel_type == t))
            .collect()
    }

    /// Find the File node ID matching `target_file`.
    pub fn file_node_id(&self) -> Option<&str> {
        self.nodes.values().find_map(|node| {
            if node.label != "File" {
                return None;
            }
            let file_path = node.properties.get("filePath")?.as_str()?;
            let matches = file_path == self.target_file
                || file_path.ends_with(&format!("/{}", self.target_file))
                || self.target_file.ends_with(&format!("/{file_path}"));
            matches.then_some(node.id.as_str())
        })
    }

    /// IDs of all symbols directly contained in a file node.
    pub fn contained_symbols(&self, file_node_id: &str) -> Vec<String> {
        self.outgoing(file_node_id, Some("CONTAINS"))
            .into_iter()
            .map(|r| r.target_id.clone())
            .collect()
    }

    /// IDs of all symbols transitively reachable via CONTAINS edges.
    ///
    /// Performs a BFS down the CONTAINS tree starting from `node_id`.
    /// Cycles are handled safely via a visited set.
    pub fn all_contained_symbols(&self, node_id: &str) -> Vec<String> {
        let mut visited: std::collections::HashSet<String> = std::collections::HashSet::new();
        let mut result = Vec::new();
        let mut frontier: VecDeque<String> = self.contained_symbols(node_id).into();
        while let Some(child) = frontier.pop_front() {
            if !visited.insert(child.clone()) {
                continue;
            }
            frontier.extend(self.contained_symbols(&child));
            result.push(child);
        }
        result
    }
}

impl Representation for ModuleDependencyGraph {
    fn name(&self) -> &str {
        "mdg"
    }

    /// Feeds the COMPOSABLE generator of `H(G_qual)`.
    fn dimension(&self) -> &str {
        "composable"
    }

    fn metrics(&self) -> HashMap<String, f64> {
        let Some(file_node_id) = self.file_node_id().map(str::to_string) else {
            // No matching File node in the graph — matches the Python
            // original's default (zero coupling, neutral 0.5
            // instability) rather than omitting the keys entirely.
            return HashMap::from([
                ("mdg.coupling".to_string(), 0.0),
                ("mdg.instability".to_string(), 0.5),
                ("mdg.fan_in".to_string(), 0.0),
                ("mdg.fan_out".to_string(), 0.0),
                ("mdg.dep_depth".to_string(), 0.0),
            ]);
        };

        let mut symbol_ids: std::collections::HashSet<String> = self
            .all_contained_symbols(&file_node_id)
            .into_iter()
            .collect();
        symbol_ids.insert(file_node_id.clone());

        let coupling = calculate_coupling(self, &file_node_id, Some(&symbol_ids));
        let instability = calculate_instability_from_result(&coupling);
        let fan = calculate_fan_in_out(self, &file_node_id, Some(&symbol_ids));
        let dep_depth = calculate_dependency_depth(self, &file_node_id);

        HashMap::from([
            ("mdg.coupling".to_string(), coupling.total() as f64),
            ("mdg.instability".to_string(), instability),
            ("mdg.fan_in".to_string(), fan.fan_in as f64),
            ("mdg.fan_out".to_string(), fan.fan_out as f64),
            ("mdg.dep_depth".to_string(), dep_depth as f64),
        ])
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use serde_json::Value;

    fn node(id: &str, label: &str, file_path: Option<&str>) -> GraphNode {
        let mut properties = HashMap::new();
        if let Some(path) = file_path {
            properties.insert("filePath".to_string(), Value::String(path.to_string()));
        }
        GraphNode {
            id: id.to_string(),
            label: label.to_string(),
            properties,
        }
    }

    fn rel(id: &str, source: &str, target: &str, rel_type: &str) -> GraphRelationship {
        GraphRelationship {
            id: id.to_string(),
            source_id: source.to_string(),
            target_id: target.to_string(),
            rel_type: rel_type.to_string(),
            confidence: 1.0,
            reason: String::new(),
            properties: Default::default(),
        }
    }

    #[test]
    fn file_node_id_matches_by_suffix() {
        let mut graph = ModuleDependencyGraph::new("src/lib.rs");
        graph.add_node(node("f1", "File", Some("/repo/src/lib.rs")));
        assert_eq!(graph.file_node_id(), Some("f1"));
    }

    #[test]
    fn all_contained_symbols_bfs_handles_cycles() {
        let mut graph = ModuleDependencyGraph::new("x");
        graph.add_relationship(rel("r1", "a", "b", "CONTAINS"));
        graph.add_relationship(rel("r2", "b", "c", "CONTAINS"));
        graph.add_relationship(rel("r3", "c", "a", "CONTAINS")); // cycle back to a
        let mut symbols = graph.all_contained_symbols("a");
        symbols.sort();
        // The starting node itself is never marked `visited` up front (it's
        // never in its own initial frontier), so a cycle back to it — as
        // here — legitimately makes it appear once in the result. This
        // matches the Python original's `visited` set exactly, which has
        // the same property; it's not a bug in either port, just a quirk
        // of "visited tracks *popped* nodes, not the start."
        assert_eq!(
            symbols,
            vec!["a".to_string(), "b".to_string(), "c".to_string()]
        );
    }

    #[test]
    fn metrics_reports_coupling_instability_fan_and_depth() {
        let mut g = ModuleDependencyGraph::new("a.py");
        g.add_node(node("File:a.py", "File", Some("a.py")));
        g.add_node(node("File:b.py", "File", Some("b.py")));
        g.add_relationship(rel("i1", "File:a.py", "File:b.py", "IMPORTS"));

        let metrics = g.metrics();
        assert_eq!(metrics["mdg.coupling"], 1.0);
        assert_eq!(metrics["mdg.instability"], 1.0); // pure efferent
        assert_eq!(metrics["mdg.fan_in"], 0.0);
        assert_eq!(metrics["mdg.fan_out"], 0.0);
        assert_eq!(metrics["mdg.dep_depth"], 1.0);
        // Graphify is advisory-only (issue #150) and must never leak into a
        // scored Representation's metrics.
        assert!(metrics.keys().all(|k| !k.starts_with("graphify")));
    }

    #[test]
    fn metrics_defaults_when_target_file_has_no_matching_node() {
        let g = ModuleDependencyGraph::new("missing.py");
        let metrics = g.metrics();
        assert_eq!(metrics["mdg.coupling"], 0.0);
        assert_eq!(metrics["mdg.instability"], 0.5);
        assert_eq!(metrics["mdg.fan_in"], 0.0);
        assert_eq!(metrics["mdg.fan_out"], 0.0);
        assert_eq!(metrics["mdg.dep_depth"], 0.0);
    }
}
